newrelic/hooks/mlmodel_langchain.py

```diff
@@ -255,10 +257,138 @@ def wrap_similarity_search(wrapped, instance, args, kwargs):
+        trace_id = available_metadata.get("trace.id", "")
+
+        try:
+            return_val = wrapped(**run_args)
+        except Exception as exc:
+            ft.notice_error(
+                attributes={
+                    "tool_id": tool_id,
+                }
+            )
+
+            run_manager_info = getattr(transaction, "_nr_run_manager_info", {})
+            if hasattr(transaction, "_nr_run_manager_info"):
+                del transaction._nr_run_manager_info
+            run_id = run_manager_info.get("run_id", "")
+
+            # Make sure the builtin attributes take precedence over metadata attributes.
+            error_tool_event_dict = {"metadata.%s" % key: value for key, value in metadata.items()} if metadata else {}
+            error_tool_event_dict.update(
+                {
+                    "id": tool_id,
+                    "run_id": run_id,
+                    "appName": app_name,
+                    "name": tool_name,
+                    "description": tool_description,
+                    "span_id": span_id,
+                    "trace_id": trace_id,
+                    "transaction_id": transaction.guid,
+                    "input": tool_input,
+                    "vendor": "langchain",
+                    "ingest_source": "Python",
+                    "duration": ft.duration,
+                    "tags": tags or "",
+                    "error": True,
+                }
+            )
+
+            transaction.record_custom_event("LlmTool", error_tool_event_dict)
+
+            raise
+
+    if not return_val:
+        return return_val
+
+    response = return_val
+
+    run_manager_info = getattr(transaction, "_nr_run_manager_info", {})
+    if hasattr(transaction, "_nr_run_manager_info"):
+        del transaction._nr_run_manager_info
+    run_id = run_manager_info.get("run_id", "")
+
+    full_tool_event_dict = {"metadata.%s" % key: value for key, value in metadata.items()} if metadata else {}
+    full_tool_event_dict.update(
+        {
+            "id": tool_id,
+            "run_id": run_id,
+            "appName": app_name,
+            "output": str(response),
+            "name": tool_name,
+            "description": tool_description,
+            "span_id": span_id,
+            "trace_id": trace_id,
+            "transaction_id": transaction.guid,
+            "input": tool_input,
+            "vendor": "langchain",
+            "ingest_source": "Python",
+            "duration": ft.duration,
+            "tags": tags or "",
+        }
+    )
+
+    transaction.record_custom_event("LlmTool", full_tool_event_dict)
+
+    return return_val
+
+
+def wrap_on_tool_start(wrapped, instance, args, kwargs):
+    run_manager = wrapped(*args, **kwargs)
+    transaction = current_transaction()
+    if not transaction:
+        return run_manager
+    # Only capture the first run_id.
+    if not hasattr(transaction, "_nr_run_manager_info"):
```

Comment:
I think this is the exact same name as I used in my PR for chains so they will conflict with each other which is not good. Whichever tools or chains finishes first, it will eat this value and it won't be available after and same goes for whichever creates it first. Maybe name this `_nr_run_manager_tools_info`?

Reply:
Just and FYI here: I ended up changing the name in the chains PR to just `_nr_chain_run_id` so we shouldn't have a conflict anymore.